### ISSUE TITLE
Should we drop the parenthesis from `.targetLocales()` queries?

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#18470]
Currently Chapel domains and arrays support a [`.targetLocales()`](https://chapel-lang.org/docs/1.25/builtins/ChapelArray.html#ChapelArray.targetLocales) query which returns the locales to which the array/domain is distributed / on which it is allocated.  It's been proposed that the parenthesis should be dropped here because the query feels more like an inherent property of the domain/array rather than something to be computed and returned.

A counterargument might be that the array value returned is nontrivial, that end-users haven't really complained about the current naming, and that the change will require updates for users / us for a fairly minor change.

**Comment by ronawho** [comment]
I believe this was done in https://github.com/chapel-lang/chapel/pull/18930. @vasslitvinov can you verify?

**Comment by vasslitvinov** [comment]
This is correct. `.targetLocales` is now paren-less, as we agreed during the Domain module review.

**Comment by ronawho** [comment]
Do we have any recommendations for maintaining compatibility across 1.25 and 1.26? I don't see a way to provide `array.targetLocales` in 1.25, so I was just going to go with `targetLocales(array)`

**Comment by bradcray** [comment]
Based on Elliot's question here, and Vass's implementation work, I find myself wondering whether we did the wrong thing here, and should go back to making it paren-ful prior to 1.26.0.  Specifically, it seems as though this is a feature that doesn't have a graceful way of being maintained across multiple versions of Chapel or being deprecated easily (see https://github.com/chapel-lang/chapel/issues/18945).  At the same time, checking the notes, I don't see much debate or discussion over the use of parens or not, and I feel as though we should bring these new realizations into the mix as part of the decision process.

**Comment by ronawho** [comment]
Hmm, my original attempt to do this was going to be:

```chpl
proc targetLocales(A: []) {
  use Reflection;
  if canResolveMethod(A, "targetLocales") {
    return A.targetLocales;
  } else {
    return A.targetLocales();
  }
}

use BlockDist;
var A = newBlockArr({1..10}, int);
writeln(targetLocales(A));
```

but this doesn't work, because `canResolveMethod(A, "targetLocales")` resolves to the paren-full method. So if we keep it parenless, I need a different way to detect the change.

**Comment by bradcray** [comment]
@ronawho: Is another way of stating your point "Maybe we should have distinct ways for querying for a paren-less method vs. a 0-argument but paren-ful method in Reflection?"

Also, if `canResolveMethod()` were to only return `true` for one of these two cases, I would expect it to work for the 0-arg, paren-ful case, which would suggest flipping the two cases in your conditional to:

```chapel
  if canResolveMethod(A, "targetLocales") {
    return A.targetLocales();
  } else {
    return A.targetLocales;
  }
```

but did that not work either?  (if it returns "true" in either case, that wouldn't help of course).

**Comment by ronawho** [comment]
I didn't have a point per-say, just wanted to mention the approach I was going to try didn't work.

**Comment by vasslitvinov** [comment]
Indeed, dropping the parens seemed too minor to discuss at length. From `notes/moduleReviews/2021-06-10-domains.txt` :

> * should targetLocales be paren-less?
>   - Vass: yes, because it's a property
>   - seemingly no objections

**Comment by bradcray** [comment]
> dropping the parens seemed too minor to discuss at length

Given the pain it's resulted in, could you add it to the re-review agenda?

> just wanted to mention the approach I was going to try didn't work.

And could you clarify:  In whatever way it didn't work, reversing the bodies of the then/else statements doesn't make it work either?

**Comment by bradcray** [comment]
@ronawho: And sorry, should've asked this the first time around, but:  If you did have a way to distinguish between those cases, do you think the error would go away?  (I'm apparently too lazy to compile your code to see the error for myself)

**Comment by ronawho** [comment]
>> just wanted to mention the approach I was going to try didn't work.

>And could you clarify: In whatever way it didn't work, reversing the bodies of the then/else statements doesn't make it work either?

`canResolveMethod(A, "targetLocales")` returns true for 1.25 and 1.26 because it will resolve for both paren and non-paren methods. 

Here's a standalone test:

```chpl
config param useParen = false;

record R {
  proc test() where useParen ==  true { return 0; }
  proc test   where useParen == false { return 1; }
}

proc test(r: R) {
  use Reflection;
  if canResolveMethod(r, "test") {
    return r.test;
  } else {
    return r.test();
  }
}

var r = new R();
writeln(test(r));
```


```
$ chpl testit.chpl
$ ./testit
1

$ chpl testit.chpl -suseParen
testit.chpl:8: In function 'test':
testit.chpl:11: error: unresolved call 'R.test'
testit.chpl:4: note: this candidate did not match: R.test()
testit.chpl:11: note: because call is written without parentheses
testit.chpl:4: note: but candidate function has parentheses
```

**Comment by vasslitvinov** [comment]
Oh wait, we check Chapel's version, right? Then the compat module does not need to rely on canResolve.

My prior attempt: how about using DavidL's PRIM_RESOLVES ? The following program detects correctly paren-ful vs. paren-less methods:
```chpl
var dom: domain(1);

proc main {
  compilerWarning(__primitive("resolves", dom.targetLocales):string);
  compilerWarning(__primitive("resolves", dom.dims):string);
  compilerError("done");
}
```
OTOH 1.25 did not have PRIM_RESOVLES, right? Disclaimer: PRIM_RESOLVES needs to be used with care.

**Comment by ronawho** [comment]
I do not want to use a primitive in Arkouda.

If we're keeping the parenless change, I can check the chapel version in the Makefile and bring in different compat modules. I did not think I'd need it in this case, but was mistaken.

**Comment by vasslitvinov** [comment]
Does Arkouda need to be backward-compatible with Chapel that's old enough that there was no `Version` module? Otherwise the version can be checked in Chapel code.

**Comment by ronawho** [comment]
Oh, yeah using the version module should work in this case since we don't need to change `import`/`use` statements. I must be doing something wrong with that attempt though:


```chpl
proc targetLocales(A: []) {
  use Version;
  if chplVersion < createVersion(1,26) {
    compilerWarning(chplVersion: string);
    return A.targetLocales();
  } else {
    return A.targetLocales;
  }
}

use BlockDist;
var A = newBlockArr({1..10}, int);
writeln(targetLocales(A));
```

```
chapel (main) $ chpl --version
chpl version 1.26.0 pre-release (0e8b99fc1c)

chapel (main) $ chpl testit.chpl
testit.chpl:13: warning: version 1.26.0 (0e8b99fc1c)
testit.chpl:1: In function 'targetLocales':
testit.chpl:5: warning: 'targetLocales' method on arrays, domains, and distributions is now paren-less; please invoke it without parentheses
```

(i.e. even with a 1.26 it's still calling the `targetLocales()` version. Seems unlikely that there's a bug in the version comparison so I must be doing something silly. Gotta head out for tonight, but will look into what I'm doing wrong tomorrow.)

**Comment by vasslitvinov** [comment]
@ronawho - our version comparison is not intuitive.
For example, `1.26.0 (0e8b99fc1c)` is actually smaller than `1.26.0`.
How about the comparison `chplVersion <= createVersion(1,25)` ?

**Comment by vasslitvinov** [comment]
A version check makes my `targetLocales()` work under both 1.25 and on today's main. I assume it works under 1.25.1 as well.
```chpl
proc targetLocales(ARG) {
  use Version;
  if chplVersion <= createVersion(1,25,1) {
    return ARG.targetLocales();
  } else {
    return ARG.targetLocales;
  }
}

use BlockDist;
var A = newBlockArr({1..10}, int);
writeln(targetLocales(A));
```

**Comment by bradcray** [comment]
@vasslitvinov : I'm remembering this morning that we failed to discuss this in last week's re-review.  I think we really need some more discussion on whether we feel strongly enough about dropping the parens here to be worth the hassle to end-users or whether people were so ambivalent about it to begin with that we should just revert to the 1.25 behavior before the 1.26 release goes out.  Can you prioritize making this discussion happen, potentially in a post-meeting discussion or by explaining the two perspectives and polling people individually?

**Comment by mppf** [comment]
While it would be OK with me to migrate from `.targetLocales()` to `.targetLocales`, this change doesn't seem to me to be worth the disruption it will cause. In other words, I don't see that we gain much by dropping the parens.
